### PR TITLE
add webhook interceptors [GH-5]

### PR DESCRIPTION
**Pull request by nhomble**
In ngbalk's [issue](https://github.com/wiremock/wiremock-webhooks-extension/issues/5), dynamically adding a date or an id to the webhook body would be great. This gets trickiers because the id I may want to post back is the same id the initial request came with. This means I'd want to have access to the ```ServeEvent``` object for more context. Thinking ahead, we may also want to add an Authorization header. So to handle all of these varying use cases, I think it's easiest to expose an interceptor for folks to have a more dynamic webhook.

### CODE REVIEW

File: src/main/java/org/wiremock/webhooks/Webhooks.java [hunk 1]
```diff
@@ -1,52 +1,76 @@
...
+        final WebhookDefinition initialDefinition = parameters.as(WebhookDefinition.class);
         final Notifier notifier = notifier();
+        final List<WebhookInterceptor> runnableInterceptors = ImmutableList.copyOf(interceptors);
+        final ServeEvent servedEvent = serveEvent;
 
         scheduler.schedule(
             new Runnable() {
                 @Override
                 public void run() {
+                    WebhookDefinition definition = initialDefinition;
```
Comment (ngbalk) [inline src/main/java/org/wiremock/webhooks/Webhooks.java:69]: Any reason for this additional assignment?

File: src/main/java/org/wiremock/webhooks/Webhooks.java [hunk 2]
```diff
@@ -1,52 +1,76 @@
...
         httpClient = HttpClientFactory.createClient();
     }
 
+    public Webhooks addInterceptor(WebhookInterceptor webhookInterceptor){
+        interceptors.add(webhookInterceptor);
+        return this;
+    }
+
+    public List<WebhookInterceptor> getInterceptors(){
+        return interceptors;
```
Comment (ngbalk) [inline src/main/java/org/wiremock/webhooks/Webhooks.java]: Immutable?

File: src/main/java/org/wiremock/webhooks/Webhooks.java (hunk 1 above)
Comment (nhomble) [reply src/main/java/org/wiremock/webhooks/Webhooks.java:69]: It's a final variable to start so that I can put it into the runnable and then I mutate the local variable in my loop

File: src/main/java/org/wiremock/webhooks/Webhooks.java [hunk 3]
```diff
@@ -1,52 +1,76 @@
...
+    }
+
+    public Webhooks setInterceptors(List<WebhookInterceptor> interceptors){
+        this.interceptors = interceptors;
```
Comment (ngbalk) [inline src/main/java/org/wiremock/webhooks/Webhooks.java]: Should we copy to a new list so its immutable from outside reference?

**Comment by nhomble** [comment]
Hey @tomakehurst any concerns with the change?

**Comment by tomakehurst** [comment]
This is a great start.

A few changes I'd suggest:
* Call it `WebhookDefinitionTransformer` rather than `WebhookInterceptor` for consistency with WireMock itself (since this is analogous to `ResponseDefinitionTransformer`)
* Support addition of the interceptor/transformer classes via an overloaded constructor, rather than the builder-type approach (thus avoiding immutability gymnastics).
* Add an extra acceptance test or two for negative cases e.g. what happens when the transformer throws an exception?

Also, I've been thinking it would be great to have (like in WM itself) templating implemented as a built-in transformer. That way it'd be possible to avoid writing Java code in a lot of cases, and use it standalone without having to bake your own JAR.

What do you think? Would you be up for adding this too?

Was thinking you could implement it very similarly, passing the request model into the template, and also a similarly structured response model.

**Comment by nhomble** [comment]
I absolutely agree on all your points @tomakehurst. I was going to have a follow up pr to provide an implementation of a *transformer* to do the templating but I am perfectly happy to do both ~now.~

Edit
I'll add issues to cover:
- matching an Http method to a URI to point to a WebhookDefinition
- including handlebar templating into the request -> WebhookDefinition configuration
and I'll aim to make some progress over the weekend

**Comment by tomakehurst** [comment]
I've pushed a release now (v1.0.0) so that should be on Maven Central in a few hours.
